test/reset/main.cpp

```diff
@@ -0,0 +1,151 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* @file main.cpp
+ *
+ * This test suite tests scenarios to reset target in-between the test.
+ *
+ *   The following tests are implemented in this file:
+ *     Test 1            Reset target inbetween the tests.
+ */
+
+#include "greentea-client/test_env.h"
+#include "utest/utest.h"
+#include "unity/unity.h"
+
+#include "mbed-drivers/mbed.h"
+
+#include <algorithm>
+#include <cstdlib>
+
+using namespace utest::v1;
+
+/** Custom echo test
+ * 
+ *  This function sends key,value pairs like {{echo;n}}. Where n is a number.
+ *  Host replies back with {{echo;m*n}}. Where m is a multiplier both target
+ *  and host know based on the target state.
+ *
+ *        state     multiplier
+ *          0           1
+ *          1           2
+ *
+ *  Response is asserted to verify the state based execution of both host
+ *  and target before and after reset.
+ */
+void do_echo_test(int response_multiplier)
+{
+    char key[] = "echo";
+    char data[] = "10";
+
+    for (int i =0; i < 10; i++)
+    {
+        snprintf(data, sizeof(data), "%d", i);
+        greentea_send_kv("echo", data);
+
+        TEST_ASSERT_NOT_EQUAL_MESSAGE(0, greentea_parse_kv(key, data, sizeof(key), sizeof(data)), "MBED: Failed to recv/parse key \"state\" value from host test!");
+        TEST_ASSERT_EQUAL_STRING("echo", key);
+
+        uint8_t count;
+        sscanf(data, "%hhu", &count);
+
+        TEST_ASSERT_EQUAL_MESSAGE(i * response_multiplier, count, "MBED: Failed to verify count received in echo!");
+    }
+}
+
+
+/** Reset test
+ * 
+ *  This function executes the reset test. It kicks off the host test by sedning
+ *  {{start;}}
+ *  Then receives state from host {{state;x}}. Where it should be x=0 on first start and 1
+ *  after reset.
+ *  In both before and after iterations it runs an echo test. It verifies the echoes based
+ *  on a multiplier known to both target and host test code based on the target state.
+ *
+ *        state     multiplier
+ *          0           1
+ *          1           2
+ *
+ */
+void reset_test_cb()
+{
+    // Start the test
+    greentea_send_kv("start", " ");
+
+    // A startup read state from the host
+    char key[] = "state";
+    char data[] = "10";
+    TEST_ASSERT_NOT_EQUAL_MESSAGE(0, greentea_parse_kv(key, data, sizeof(key), sizeof(data)), "MBED: Failed to recv/parse key \"state\" value from host test!");
+
+    uint8_t state;
+    sscanf(data, "%hhu", &state);
+    
+    TEST_ASSERT_TRUE(state <= 1);
+
+    // Start echo test with a response multiplier that is shared secret between target and host to demonstrated
+    // state correspondance. 
+    // state   multiplier
+    //   0        1
+    //   1        2
+
+    if (state == 0)
+    {
+        do_echo_test(1);
+
+        // change state to 1 and reset.
+        greentea_send_kv("state", "1");
+        greentea_send_kv("reset", " ");
```

Comment:
It could be better (?) to acutally send some payload to make it more verbose. It is designed to be a key-value thing so we send pairs that together make sense like ```reset```;```now``` or ```reset```;```software_reset```` or ```timeout```;```10``` (secs) etc...
```c++
greentea_send_kv("reset", "software_reset");
```